drivers/wiegand/wiegand.c

```diff
@@ -0,0 +1,298 @@
+    dev->wg_card_temp_high |= ((0x80000000 & dev->wg_card_temp)>>31);
+    dev->wg_card_temp_high <<= 1;
+    dev->wg_card_temp <<=1;
+  }
+  else
+  {
+    // D0 represent binary 0, so just left shift card data
+    dev->wg_card_temp <<= 1;
+  }
+  // Keep track of last wiegand bit received
+  dev->wg_last_wiegand = xtimer_now_usec64();
+}
+
+/**
+ * @brief Handler for D1 signal in Wiegand protocol
+ *
+ */
+void wg_read_d1(void *arg)
+{
+  wiegand_t *dev = (wiegand_t *)arg;
+
+  // Increment bit count for Interrupt connected to D1
+  dev->wg_bit_count ++;
+
+  // If bit count more than 31, process high bits
+  if (dev->wg_bit_count>31)
+  {
+    // shift value to high bits
+    dev->wg_card_temp_high |= ((0x80000000 & dev->wg_card_temp)>>31);
+    dev->wg_card_temp_high <<= 1;
+    dev->wg_card_temp |= 1;
+    dev->wg_card_temp <<=1;
+  }
+  else
+  {
+    // D1 represent binary 1, so OR card data with 1 then
+    dev->wg_card_temp |= 1;
+    // left shift card data
+    dev->wg_card_temp <<= 1;
+  }
+  // Keep track of last wiegand bit received
+  dev->wg_last_wiegand = xtimer_now_usec64();
+}
+
+/**
+ * @brief Returns the last code read from the Wiegand device.
+ *
+ */
+unsigned long wg_get_code(wiegand_t *dev)
+{
+  return dev->code;
+}
+
+/**
+ * @brief Returns the Wiegand type.
+ *
+ */
+int16_t wg_get_wiegand_type(wiegand_t *dev)
+{
+  return dev->type;
+}
+
+/**
+ * @brief Translates * and # keys to [ENTER] and [ESCAPE] keys.
+ *
+ */
+char translate_enter_escape_key_press(char originalKeyPress) {
+  switch(originalKeyPress) {
+    case 0x0b:     // 11 or * key
+      return 0x0d; // 13 or ASCII ENTER
+
+    case 0x0a:     // 10 or # key
+      return 0x1b; // 27 or ASCII ESCAPE
+
+    default:
+      return originalKeyPress;
+  }
+}
+
+/**
+ * @brief Disables interrupt handling for D0 and D1 lines.
+ *
+ */
+static inline void wg_disable_int(wiegand_t *dev)
+{
+  gpio_irq_disable(dev->conn->d0);
+  gpio_irq_disable(dev->conn->d1);
+}
+
+/**
+ * @brief Enables interrupt handling for D0 and D1 lines.
+ *
+ */
+static inline void wg_enable_int(wiegand_t *dev)
+{
+  gpio_irq_enable(dev->conn->d0);
+  gpio_irq_enable(dev->conn->d1);
+}
+
+/**
+ * @brief Returns the card identification.
+ *
+ */
+unsigned long wg_get_card_id(volatile unsigned long *codehigh, volatile unsigned long *codelow, char bitlength)
+{
+
+  if (bitlength==26) // EM tag
+    return (*codelow & 0x1FFFFFE) >>1;
+
+  if (bitlength==34) // Mifare
+  {
+    *codehigh = *codehigh & 0x03; // only need the 2 LSB of the codehigh
+    *codehigh <<= 30; // shift 2 LSB to MSB
+    *codelow >>=1;
+    return *codehigh | *codelow;
+  }
+  return *codelow; // EM tag or Mifare without parity bits
+}
+
+/**
+ * @brief Process the temporary Wiegand buffer and decodes the data.
+ *
+ */
+bool wg_do_conversion(wiegand_t *dev)
+{
+  uint64_t sysTick;
+
+  sysTick = xtimer_now_usec64();
+
+  if ((sysTick - dev->wg_last_wiegand) > 25000) // if no more signal coming through after 25ms
```

Comment:
Instead you could do `thread_sleep()` in `wg_get_code()` and call [`xtimer_set_wakeup(timer, timeout, thread_getpid())`](https://riot-os.org/api/group__sys__xtimer.html#gab46b49ec5cf112476c83ca727bb77f67) before.

Then instead of `dev->wg_last_wiegand = xtimer_now_usec64();` you can just call `xtimer_set(timer, 25000);` to set the timeout.

Reply:
I didn't really wanted to make any function blocking because an RFID reader could stay without any activity during several hours. I know that the sleep strategy essentially will not use any resources but it's more just a design approach to keep the usage more readable and easy to integrate.

May be I can provide a blocking function with a timeout specified by the user.